applications/FemToDemApplication/custom_processes/generate_initial_skin_DEM_process.cpp

```diff
@@ -0,0 +1,152 @@
+            Vector distances(r_neigh_nodes.size());
+    
+            // Loop over the neighbours of that node
+            bool has_dem_neigh = false;
+            for (int neigh = 0; neigh < r_neigh_nodes.size(); neigh++) {
+                auto& r_neighbour_node = r_neigh_nodes[neigh];
+                const double dist_between_nodes = CalculateDistanceBetweenNodes(it_node, r_neighbour_node);
+                distances(neigh) = dist_between_nodes;
+                if (r_neighbour_node.GetValue(IS_DEM)) { // Has DEM
+                    has_dem_neigh = true;
+                    potential_radii(neigh) = dist_between_nodes - r_neighbour_node.GetValue(RADIUS);
+                    if (potential_radii(neigh) < 0.0 || potential_radii(neigh) / dist_between_nodes < 0.2) {
+                        const double new_radius = dist_between_nodes*0.5;
+                        auto& pDEM_particle = r_neighbour_node.GetValue(DEM_PARTICLE_POINTER);
+                        auto& r_radius_neigh_old = pDEM_particle->GetGeometry()[0].GetSolutionStepValue(RADIUS);
+                        r_radius_neigh_old = new_radius;
+                        r_neighbour_node.SetValue(RADIUS, new_radius);
+                        potential_radii(neigh) = new_radius;
+                    }
+                } else {
+                    potential_radii(neigh) = 0.0;
+                }
+            }
+            // Let's compute the Radius of the new DEM
+            double radius;
+            if (has_dem_neigh) {
+                radius = this->GetMinimumValue(potential_radii);
+            } else {
+                radius = this->GetMinimumValue(distances)*0.5;
+            }
+            const array_1d<double,3>& r_coordinates = it_node->Coordinates();
+            const int id = this->GetMaximumDEMId() + 1;
+            this->CreateDEMParticle(id, r_coordinates, p_DEM_properties, radius, it_node);
+            num_DEM++;
+        }
+    }
+}
+
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void GenerateInitialSkinDEMProcess::CreateDEMParticle(
+    const int Id,
+    const array_1d<double, 3> Coordinates,
+    const Properties::Pointer pProperties,
+    const double Radius,
+	NodeIteratorType rNode
+)
+{
+    auto spheric_particle = mParticleCreator.CreateSphericParticleRaw(mrDEMModelPart, Id, Coordinates, pProperties, Radius, "SphericParticle3D");
+    rNode->SetValue(IS_DEM, true);
+    rNode->SetValue(RADIUS, Radius);
+    rNode->SetValue(DEM_PARTICLE_POINTER, spheric_particle);
+}
+
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double GenerateInitialSkinDEMProcess::CalculateDistanceBetweenNodes(
+    NodeIteratorType rNode1, 
+    const NodeType& rNode2
+    )
+{
+    const double X1 = rNode1->X();
```

Comment:
I would recommend to replace this with COrrdinates and avoid acccess to the database several times and compute directly the norm of the difference